quickwit/quickwit-ingest/src/ingest_v2/fetch.rs

```diff
@@ -423,79 +359,215 @@ async fn fault_tolerant_fetch_task(
     index_uid: IndexUid,
     source_id: SourceId,
     shard_id: ShardId,
-    mut from_position_exclusive: Option<u64>,
-    to_position_inclusive: Option<u64>,
-
-    preferred_ingester_id: NodeId,
-    mut failover_ingester_id: Option<NodeId>,
-
+    mut from_position_exclusive: Position,
+    to_position_inclusive: Position,
+    ingester_ids: Vec<NodeId>,
     ingester_pool: IngesterPool,
-
-    mut fetch_stream: ServiceStream<IngestV2Result<FetchResponseV2>>,
     fetch_response_tx: mpsc::Sender<IngestV2Result<FetchResponseV2>>,
 ) {
-    while let Some(fetch_response_result) = fetch_stream.next().await {
-        match fetch_response_result {
-            Ok(fetch_response) => {
-                from_position_exclusive = fetch_response.to_position_inclusive();
-                if fetch_response_tx.send(Ok(fetch_response)).await.is_err() {
-                    // The stream was dropped.
-                    break;
+    'outer: for (ingester_idx, ingester_id) in ingester_ids.iter().enumerate() {
```

Comment:
I suspect this piece is needlessly convoluted. we should probably try to simplify it...

Reply:
Added a TODO. We could break the logic into smaller functions.